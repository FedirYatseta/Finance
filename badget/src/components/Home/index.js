import React, { useState, useEffect } from 'react'
import Balance from '../Balance';
import ErrorBoundary from '../ErrorBoundry';
import Form from '../Form';
import Transactions from '../Transactions';

import { addItem, getItems } from '../../utils/indexdb'

const Home = () => {

    const [balance, setBalance] = useState(0);
    const [transactions, setTransaction] = useState([])
    useEffect(() => {
        getItems().then((item) => {
            setTransaction(item)
        }).catch((e) => {
            console.error('error', e)
        })
    }, [setTransaction])


    const onChange = ({ value, date, comment }) => {

        const transaction = {
            value: +value,
            date,
            comment,
            id: Date.now()
        }

        setTransaction([transaction
            , ...transactions])
        setBalance(balance + Number(value))

        addItem(transaction);
    }

    return (
        <>
            <ErrorBoundary >
                <Balance balance={balance} />
                <Form onChange={onChange} />
                <hr />
                <Transactions transactions={transactions} />
            </ErrorBoundary>

        </>
    )
}


export default Home;