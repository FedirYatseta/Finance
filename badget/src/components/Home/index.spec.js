//import { shallow } from 'enzyme';
import { mount } from 'enzyme';
import React from 'react';
import Home from '.';
import { addItem, getItems } from '../../utils/indexdb'
import { act } from 'react-dom/test-utils'

jest.mock('../../utils/indexdb', () => ({
  addItem: jest.fn(),
  getItems: jest.fn()
}))


describe('Home', () => {
  let sut;
  let props;

  describe("when component is mounted", () => {


    describe("when transaction are returned successfully", () => {
      beforeEach(() => {
        getItems.mockImplementation(() => Promise.resolve([{ value: 1 }]))
      })
      beforeEach(async () => {
        await act(async () => {
          sut = mount(<Home {...props} />)
        });
        sut.update();

      })
      it('it should zero', () => {
        const { balance } = sut.find('Balance').at(0).props();
        expect(balance).toBe(0)
      }
      )

      it("should render transaction with one item", () => {

        const { transactions } = sut.find("Transactions").at(0).props();
        expect(transactions).toEqual([{ value: 1 }])
      })
    })

    describe("when transaction are returned error", () => {
      let consoleSpy;
      beforeEach(() => {
        consoleSpy = jest.spyOn(console, "error")
        getItems.mockImplementation(() => Promise.reject("My Error"))
      })
      beforeEach(async () => {
        await act(async () => {
          sut = mount(<Home {...props} />)
        });
        sut.update();

      })


      it("should render transaction with error", () => {

        const { transactions } = sut.find("Transactions").at(0).props();
        expect(transactions).toEqual([])
      })
      it("should consol error reject message ", () => {

        expect(consoleSpy).toHaveBeenCalledWith('error', "My Error")
      })
    })

    describe("when transaction is added ", () => {
      it('it should increase balance', () => {
        const { onChange } = sut.find("Form").at(0).props();
        onChange({
          value: 1,
          date: '',
          comment: '',
        })
        sut.update();

        const { balance } = sut.find("Balance").at(0).props();
        expect(balance).toBe(1)
      })



    })
  })

})