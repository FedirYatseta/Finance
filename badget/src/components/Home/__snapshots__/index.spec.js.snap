// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should show Home 1`] = `
Array [
  <div>
    <div
      className="sc-bczRLJ cKPmbC"
      value={0}
    >
      Balance: 
      0
    </div>
  </div>,
  <form
    className="sc-dkzDqf esGgqa"
    onSubmit={[Function]}
  >
    <input
      className="sc-hKMtZM chszIM"
      name="date"
      onChange={[Function]}
      type="date"
      value="2022-04-13"
    />
    <textarea
      className="sc-eCYdqJ gqnvUR"
      name="comment"
      onChange={[Function]}
      type="text"
      value=""
    />
    <input
      className="sc-hKMtZM chszIM"
      name="value"
      onChange={[Function]}
      placeholder="Sum"
      type="number"
      value=""
    />
    <button
      className="sc-jSMfEi btHlbY"
    >
      Save
    </button>
  </form>,
  <hr />,
  <div>
    <hr />
  </div>,
]
`;
